cli/module_reload.go

```diff
@@ -0,0 +1,109 @@
+package cli
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	apppb "go.viam.com/api/app/v1"
+
+	rdkConfig "go.viam.com/rdk/config"
+	"go.viam.com/rdk/logging"
+)
+
+// ModuleMap is a type alias to indicate where a map represents a module config.
+// We don't convert to rdkConfig.Module because it can get out of date with what's in the db.
+// Using maps directly also saves a lot of high-maintenance ser/des work.
+type ModuleMap map[string]interface{}
+
+// configureModule is the configuration step of module reloading. Returns (needsRestart, error).
+func configureModule(vc *viamClient, manifest *moduleManifest, part *apppb.RobotPart) (bool, error) {
+	logger := logging.Global()
+	if manifest == nil {
+		return false, fmt.Errorf("reconfiguration requires valid manifest json passed to --%s", moduleFlagPath)
+	}
+	partMap := part.RobotConfig.AsMap()
+	modules, err := mapOver(
+		partMap["modules"].([]interface{}),
+		func(raw interface{}) (ModuleMap, error) { return ModuleMap(raw.(map[string]interface{})), nil },
+	)
+	if err != nil {
+		return false, err
+	}
+
+	modules, dirty, err := mutateModuleConfig(modules, *manifest)
+	if err != nil {
+		return false, err
+	}
+	// note: converting to interface{} or else proto serializer will fail downstream in NewStruct.
+	modulesAsInterfaces, err := mapOver(modules, func(mod ModuleMap) (interface{}, error) {
+		return map[string]interface{}(mod), nil
+	})
+	if err != nil {
+		return false, err
+	}
+	partMap["modules"] = modulesAsInterfaces
+	if dirty {
+		logger.Debug("writing back config changes")
+		err = vc.updateRobotPart(part, partMap)
+		if err != nil {
+			return false, err
+		}
+	}
+	// if we modified config, caller doesn't need to restart module.
+	return !dirty, nil
+}
+
+// localizeModuleID converts a module ID to its 'local mode' name.
+// TODO(APP-4019): remove this logic after registry modules can have local ExecPath.
+func localizeModuleID(moduleID string) string {
+	return "hr_" + strings.ReplaceAll(moduleID, ":", "_")
```

Comment:
Could you briefly explain what mutation this is doing? Prefixing with `hr_` and replacing `:` with `_`? What does the `hr` prefix mean?

Reply:
hr for 'hot reloading' -- open to other / clearer prefixes. purpose is to make collisions less likely

this function takes a module ID like `viam:my-module` and converts it to a name like `hr_viam_my-module`

couple of things happening here:
- the user wants to test a local build of a module with id=x. we need to remove or replace that module with the local one or else the models will collide
- I'm converting Module.ModuleID -> Module.Name because type=registry modules can't have an executable_path. (validation [here](https://github.com/viamrobotics/app/blob/28d2ef52/robotservice/config_retriever.go#L689), but this changes after APP-4019)
- removing `:` because of [ModuleNameValidator regex](https://github.com/viamrobotics/app/blob/28d2ef52/data/module.go#L46)

Reply:
because this is user-facing (it mutates their config), I think it is important that the user understands why a mutation occured, and I think a verbose prefix like  `hot_reloading_` is better.

Reply:
f284633b changed hr_ prefix to _from_reloading suffix. ('reloading' prefix made it sound like the module is in a reloading state)